staging/src/k8s.io/apiserver/pkg/endpoints/installer.go

```diff
@@ -994,6 +994,9 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 	if categoriesProvider, ok := storage.(rest.CategoriesProvider); ok {
 		apiResource.Categories = categoriesProvider.Categories()
 	}
+	if singularNameProvider, ok := storage.(rest.SingularNameProvider); ok {
```

Comment:
this was the bit that we were saying should be non-optional... storage *must* implement this

Reply:
having an interface is fine, just error if the storage doesn't implement the interface and we're going to surface any endpoints for it (`len(actions) > 0`)

Reply:
it makes sense now thanks.